In [1]:
import pandas as pd
import numpy as np
from pathlib import Path
import os

ROOT_FOLDER = "NYPL-menus"
# ROOT_FOLDER = "NYPL-menus-cleaned"

OUTPUT_FOLDER = "NYPL-menus-cleaned"
os.makedirs(OUTPUT_FOLDER, exist_ok=True)

OUTPUT_FILE = []

MENU = 1
MENU_PAGE = 3
MENU_ITEM = 2
DISH = 0

# Data Cleaning Workflow Instructions

## Accessing Datasets

Each dataset can be accessed from the dataset collection using the following keys:

```python
dataset["MENU"]          # Menu data
dataset["MENU_PAGE"]     # Menu page data
dataset["MENU_ITEM"]     # Menu item data 
dataset["DISH"]          # Dish data
```

## Cleaning the Cleaned Dataset

Update the ROOT_FOLDER variable to point to your cleaned data directory:

```python
ROOT_FOLDER = "NYPL-menus-cleaned"  # Path to cleaned data
```

In [2]:
dataset = []

print(f"Loading datasets from {ROOT_FOLDER}...")
for filename in sorted(Path(ROOT_FOLDER).iterdir()):
    print(f"Loading {filename.name}...")
    OUTPUT_FILE.append(filename.name.replace(".csv", "_fixed.csv"))
    if filename.name.endswith(".csv"):
        dataset.append(pd.read_csv(filename, na_values=[""]))

Loading datasets from NYPL-menus...
Loading Dish.csv...
Loading Menu.csv...
Loading MenuItem.csv...
Loading MenuPage.csv...


In [3]:
# IC 1: Temporal consistency in Dish
ic1_violations = dataset[DISH][dataset[DISH]["first_appeared"] > dataset[DISH]["last_appeared"]]

print(f"Violations found: {len(ic1_violations)}")
ic1_violations.head(10)

Violations found: 6


id                                               name  \
131193  164029                                   Clear beef broth   
163257  204888                          Hot roast beef with gravy   
197050  250693  SURI LEBERLI - Shredded Calf's Liver Flambe in...   
197053  250699                          SWISS MINCED VEAL, ROESTI   
237740  301736                                        Cafe Glacee   
244534  309629                                      Garlic Butter   

        description  menus_appeared  times_appeared  first_appeared  \
131193          NaN               0               1            1900   
163257          NaN               0               1            1900   
197050          NaN               0               1            1945   
197053          NaN               0               1            1945   
237740          NaN               0               2            1940   
244534          NaN               0               1            1947   

        last_appeared  lowest_price  highest_price  
131193              0          0.25           0.25  
163257              0          0.25           0.25  
197050              0           NaN            NaN  
197053              0           NaN            NaN  
237740              0          0.40           0.40  
244534              0          0.40           0.40

In [4]:
# IC 1: Temporal consistency in Dish cleaning
dataset[DISH].loc[ic1_violations.index, "last_appeared"] = dataset[DISH].loc[ic1_violations.index, "first_appeared"]

print(f"After Cleaning Applied: {len(ic1_violations)}")
dataset[DISH].loc[ic1_violations.index]

After Cleaning Applied: 6


id                                               name  \
131193  164029                                   Clear beef broth   
163257  204888                          Hot roast beef with gravy   
197050  250693  SURI LEBERLI - Shredded Calf's Liver Flambe in...   
197053  250699                          SWISS MINCED VEAL, ROESTI   
237740  301736                                        Cafe Glacee   
244534  309629                                      Garlic Butter   

        description  menus_appeared  times_appeared  first_appeared  \
131193          NaN               0               1            1900   
163257          NaN               0               1            1900   
197050          NaN               0               1            1945   
197053          NaN               0               1            1945   
237740          NaN               0               2            1940   
244534          NaN               0               1            1947   

        last_appeared  lowest_price  highest_price  
131193           1900          0.25           0.25  
163257           1900          0.25           0.25  
197050           1945           NaN            NaN  
197053           1945           NaN            NaN  
237740           1940          0.40           0.40  
244534           1947          0.40           0.40

In [5]:
# IC 2: Date Outliner in Menu

# Extract first 4 digits of call_number (if not null)
dataset[MENU]["call_prefix"] = dataset[MENU]["call_number"].str[:4]

# Extract first 4 digits of date (if not null)
dataset[MENU]["date_prefix"] = dataset[MENU]["date"].str[:4]

# Check constraint: call_prefix == date_year when both exist
ic2_violations = dataset[MENU][
    dataset[MENU]["call_number"].notna() & 
    dataset[MENU]["date"].notna() & 
    dataset[MENU]["call_prefix"].str.isnumeric() &
    dataset[MENU]["date"].str.match(r"[^1]+") & 
    (dataset[MENU]["call_prefix"] != dataset[MENU]["date_prefix"])
]

print(f"Before Cleaning Applied: {len(ic2_violations)}")
dataset[MENU].loc[ic2_violations.index][["id", "call_number", "date", "call_prefix", "date_prefix"]]

Before Cleaning Applied: 4


id     call_number        date call_prefix date_prefix
525    13112       1900-2328  0190-03-06        1900        0190
14659  32265  1918-0387_wotm  2928-03-26        1918        2928
16917  34727  1912-0667_wotm  0001-01-01        1912        0001
16918  34728  1912-0668_wotm  0001-01-01        1912        0001

In [6]:

# IC 2: Date Outliner in Menu Cleaning
dataset[MENU].loc[ic2_violations.index, "date"] = (
    dataset[MENU].loc[ic2_violations.index, "call_prefix"] + 
    dataset[MENU].loc[ic2_violations.index, "date"].str[4:]
)

print(f"After Cleaning Applied: {len(ic2_violations)}")
dataset[MENU].loc[ic2_violations.index][["id", "call_number", "date", "call_prefix", "date_prefix"]]

After Cleaning Applied: 4


id     call_number        date call_prefix date_prefix
525    13112       1900-2328  1900-03-06        1900        0190
14659  32265  1918-0387_wotm  1918-03-26        1918        2928
16917  34727  1912-0667_wotm  1912-01-01        1912        0001
16918  34728  1912-0668_wotm  1912-01-01        1912        0001

In [7]:
# IC 3: Date Blank with call number year in Menu

# Extract first 4 digits of call_number (if not null)
dataset[MENU]["call_prefix"] = dataset[MENU]["call_number"].str[:4]

# Extract first 4 digits of date (if not null)
dataset[MENU]["date_prefix"] = dataset[MENU]["date"].str[:4]

# Check constraint: date should not be blank when call_prefix is numeric
ic3_violations = dataset[MENU][
    dataset[MENU]["call_number"].notna() & 
    dataset[MENU]["date"].isna() &
    dataset[MENU]["call_prefix"].str.isnumeric()
]

print(f"Violations found: {len(ic3_violations)}")
ic3_violations[["id", "call_number", "date", "call_prefix", "date_prefix"]].head(20)

Violations found: 47


id      call_number date call_prefix date_prefix
465    13042        1900-2517  NaN        1900         NaN
4598   20978         1906-783  NaN        1906         NaN
5025   21467         1886-036  NaN        1886         NaN
5400   21969         1887-028  NaN        1887         NaN
8633   25998         1900-189  NaN        1900         NaN
8732   26119         1899-606  NaN        1899         NaN
10093  27576         1910-881  NaN        1910         NaN
10426  27912   1973-0020_wotm  NaN        1973         NaN
10526  28012   1977-0004_wotm  NaN        1977         NaN
10560  28062   1978-0021_wotm  NaN        1978         NaN
10648  28151   1967-0067_wotm  NaN        1967         NaN
10649  28152   1967-0068_wotm  NaN        1967         NaN
10780  28285   1965-0097_wotm  NaN        1965         NaN
10788  28294   1983-0002_wotm  NaN        1983         NaN
10789  28296  1980-0051A_wotm  NaN        1980         NaN
10807  28314   1980-0013_wotm  NaN        1980         NaN
10842  28349   1984-0010_wotm  NaN        1984         NaN
10844  28351   1984-0012_wotm  NaN        1984         NaN
10845  28352   1984-0015_wotm  NaN        1984         NaN
10898  28405   1958-0053_wotm  NaN        1958         NaN

In [8]:
# IC 3: Date Blank with call number year in Menu Cleaning
dataset[MENU].loc[ic3_violations.index, "date"] = (
    dataset[MENU].loc[ic3_violations.index, "call_prefix"] + "-01-01"
)

print(f"After Cleaning Applied: {len(ic3_violations)}")
dataset[MENU].loc[ic3_violations.index][["id", "call_number", "date", "call_prefix", "date_prefix"]]

After Cleaning Applied: 47


id      call_number        date call_prefix date_prefix
465    13042        1900-2517  1900-01-01        1900         NaN
4598   20978         1906-783  1906-01-01        1906         NaN
5025   21467         1886-036  1886-01-01        1886         NaN
5400   21969         1887-028  1887-01-01        1887         NaN
8633   25998         1900-189  1900-01-01        1900         NaN
8732   26119         1899-606  1899-01-01        1899         NaN
10093  27576         1910-881  1910-01-01        1910         NaN
10426  27912   1973-0020_wotm  1973-01-01        1973         NaN
10526  28012   1977-0004_wotm  1977-01-01        1977         NaN
10560  28062   1978-0021_wotm  1978-01-01        1978         NaN
10648  28151   1967-0067_wotm  1967-01-01        1967         NaN
10649  28152   1967-0068_wotm  1967-01-01        1967         NaN
10780  28285   1965-0097_wotm  1965-01-01        1965         NaN
10788  28294   1983-0002_wotm  1983-01-01        1983         NaN
10789  28296  1980-0051A_wotm  1980-01-01        1980         NaN
10807  28314   1980-0013_wotm  1980-01-01        1980         NaN
10842  28349   1984-0010_wotm  1984-01-01        1984         NaN
10844  28351   1984-0012_wotm  1984-01-01        1984         NaN
10845  28352   1984-0015_wotm  1984-01-01        1984         NaN
10898  28405   1958-0053_wotm  1958-01-01        1958         NaN
10966  28474   1985-0002_wotm  1985-01-01        1985         NaN
11073  28581   1985-0017_wotm  1985-01-01        1985         NaN
11078  28586   1985-0022_wotm  1985-01-01        1985         NaN
11084  28592   1985-0028_wotm  1985-01-01        1985         NaN
11264  28775   1987-0099_wotm  1987-01-01        1987         NaN
11321  28832   1987-0065_wotm  1987-01-01        1987         NaN
11683  29202   1952-0003_wotm  1952-01-01        1952         NaN
11798  29318   2001-0018_wotm  2001-01-01        2001         NaN
12792  30337   1940-0001_wotm  1940-01-01        1940         NaN
12909  30457   1941-0004_wotm  1941-01-01        1941         NaN
13312  30863   1938-0026_wotm  1938-01-01        1938         NaN
13313  30864   1938-0027_wotm  1938-01-01        1938         NaN
13371  30923   1938-0189_wotm  1938-01-01        1938         NaN
13391  30952   1938-0193_wotm  1938-01-01        1938         NaN
13418  30994   1938-0143_wotm  1938-01-01        1938         NaN
13475  31062   1933-0136_wotm  1933-01-01        1933         NaN
13636  31223   1933-0003_wotm  1933-01-01        1933         NaN
13637  31224   1933-0004_wotm  1933-01-01        1933         NaN
13638  31225   1933-0005_wotm  1933-01-01        1933         NaN
13866  31456   1932-0025_wotm  1932-01-01        1932         NaN
13899  31489   1931-0003_wotm  1931-01-01        1931         NaN
13915  31505   1931-0020_wotm  1931-01-01        1931         NaN
14050  31642   1926-0001_wotm  1926-01-01        1926         NaN
14075  31667   1920-0009_wotm  1920-01-01        1920         NaN
14195  31790   1925-0009_wotm  1925-01-01        1925         NaN
14213  31808   1925-0027_wotm  1925-01-01        1925         NaN
14280  31876   1921-0057_wotm  1921-01-01        1921         NaN

In [9]:
# IC 4: Date Blank with no date info in call number in Menu

# Extract first 4 digits of call_number (if not null)
dataset[MENU]["call_prefix"] = dataset[MENU]["call_number"].str[:4]

# Extract first 4 digits of date (if not null)
dataset[MENU]["date_prefix"] = dataset[MENU]["date"].str[:4]

# Check constraint: date should not be blank when call_prefix is numeric
ic4_violations = dataset[MENU][
    dataset[MENU]["call_number"].notna() & 
    dataset[MENU]["date"].isna() & (
    (dataset[MENU]["call_prefix"] == 'Zand') |
    (dataset[MENU]["call_prefix"] == 'Soet') |
    (dataset[MENU]["call_prefix"] == 'soet') |
    (dataset[MENU]["call_prefix"] == 'Bara') |
    (dataset[MENU]["call_prefix"] == '_wot'))
]

print(f"Violations found: {len(ic4_violations)}")
dataset[MENU].loc[ic4_violations.index]

Violations found: 536


id name                                            sponsor  \
8913   26347  NaN                                       Pan American   
8915   26349  NaN                              American Export Lines   
8918   26352  NaN                                  American Airlines   
8921   26355  NaN                                       Pan American   
8922   26356  NaN                                       Pan American   
...      ...  ...                                                ...   
10224  27708  NaN  Dinner held by Pennsylvania Railroad (Railroad...   
10225  27709  NaN  Dinner held by Pennsylvania Railroad (Railroad...   
10226  27710  NaN               Dinner held by Pennsylvania Railroad   
10228  27712  NaN  Exec Committe California Midwinter Int Exposition   
13921  31511  NaN                                        Hotel Astor   

                  event      venue              place  \
8913                NaN    AIRLINE                NaN   
8915   Captain's Dinner  STEAMSHIP  S.S. Independence   
8918                NaN    AIRLINE                NaN   
8921                NaN    AIRLINE                NaN   
8922                NaN    AIRLINE                NaN   
...                 ...        ...                ...   
10224               NaN        NaN                NaN   
10225               NaN        NaN                NaN   
10226               NaN        NaN                NaN   
10228               NaN        NaN                NaN   
13921               NaN        NaN                NaN   

                                    physical_description occasion  \
8913                              Folder; 8.5 x 6 inches      NaN   
8915                          Booklet; 8.5 x 11.5 inches      NaN   
8918   Folded set of postcards; 7 x 4.5 in. folded; 7...      NaN   
8921                               Folder; 11 x 9 inches      NaN   
8922   Tri-fold; 9.75 x 13.25 in. folded; 9.75 x 26 i...      NaN   
...                                                  ...      ...   
10224                            10 x 7 in. fold. ; ill.      NaN   
10225                            10 x 7 in. fold. ; ill.      NaN   
10226                                                NaN      NaN   
10228                                                NaN      NaN   
13921                                                NaN      NaN   

                                                   notes        call_number  \
8913   Dieter Zander Collection; cover is an illustra...          Zander 11   
8915   Dieter Zander Collection; captain, officers, a...          Zander 13   
8918   Dieter Zander Collection; Americana themed; sc...          Zander 16   
8921   Dieter Zander Collection; cover is a watercolo...  Zander 19 undated   
8922   Dieter Zander Collection, cover is a watercolo...  Zander 21 undated   
...                                                  ...                ...   
10224                                                NaN       Soete 235A-B   
10225                                                NaN       Soete 237A-C   
10226                                                NaN         Soete 209A   
10228                                                NaN           soete 56   
13921                                                NaN              _wotm   

       ...  date                                           location  \
8913   ...   NaN                                       Pan American   
8915   ...   NaN                              American Export Lines   
8918   ...   NaN                                  American Airlines   
8921   ...   NaN                                       Pan American   
8922   ...   NaN                                       Pan American   
...    ...   ...                                                ...   
10224  ...   NaN                              Pennsylvania Railroad   
10225  ...   NaN                              Pennsylvania Railroad   
10226  ...   NaN               Dinner Held 

In [10]:

# IC 4: Date Blank with no date info in call number in Menu Cleaning
dataset[MENU]["date"] = dataset[MENU]["date"].ffill().bfill().interpolate(method='nearest')

print(f"After Cleaning Applied: {len(ic4_violations)}")
dataset[MENU].loc[ic4_violations.index][["id", "call_number", "date", "call_prefix", "date_prefix"]]

After Cleaning Applied: 536


C:\Users\Hwu55\AppData\Local\Temp\ipykernel_6412\1732838106.py:2: FutureWarning: Series.interpolate with object dtype is deprecated and will raise in a future version. Call obj.infer_objects(copy=False) before interpolating instead.
  dataset[MENU]["date"] = dataset[MENU]["date"].ffill().bfill().interpolate(method='nearest')


id        call_number        date call_prefix date_prefix
8913   26347          Zander 11  1989-01-01        Zand         NaN
8915   26349          Zander 13  1969-12-01        Zand         NaN
8918   26352          Zander 16  1952-01-01        Zand         NaN
8921   26355  Zander 19 undated  1963-07-29        Zand         NaN
8922   26356  Zander 21 undated  1963-07-29        Zand         NaN
...      ...                ...         ...         ...         ...
10224  27708       Soete 235A-B  1919-05-12        Soet         NaN
10225  27709       Soete 237A-C  1919-05-12        Soet         NaN
10226  27710         Soete 209A  1919-05-12        Soet         NaN
10228  27712           soete 56  1931-11-01        soet         NaN
13921  31511              _wotm  1931-09-18        _wot         NaN

[536 rows x 5 columns]

In [ ]:
# IC 5: Date range outsite of 1890-1970 in Menu
ic5_violations = dataset[MENU][
    dataset[MENU]["date"].notna() & 
    (dataset[MENU]["date"].str[:4].astype(int) < 1890) | 
    (dataset[MENU]["date"].str[:4].astype(int) > 1970)
]["id"]

print(f"Violations found: {len(ic5_violations)}")
print(f"Menu dataset size: {len(dataset[MENU])}")
print(f"MenuPage dataset size: {len(dataset[MENU_PAGE])}")
print(f"MenuItem dataset size: {len(dataset[MENU_ITEM])}")
dataset[MENU].loc[ic5_violations.index]

Violations found: 1248
Menu dataset size: 17545
MenuPage dataset size: 66937
MenuItem dataset size: 1332726
Dish dataset size: 423397


id                                               name  \
39     12503                                                NaN   
48     12515                                                NaN   
49     12516                                                NaN   
147    12635                                                NaN   
148    12636                                                NaN   
...      ...                                                ...   
11875  29396                                                NaN   
11877  29398                                                NaN   
11878  29399                                                NaN   
11880  29401                                                NaN   
16425  34079  New Map of Brooklyn and vicinity, for the Broo...   

                                                 sponsor  \
39             POLICE DEPARTMENT OF THE CITY OF NEW YORK   
48                                            THE ALBANY   
49                                          REVERE HOUSE   
147                                                  NaN   
148                                                  NaN   
...                                                  ...   
11875                                 Hotel Mas De Vence   
11877                                            Outpost   
11878                                        Que Pasta's   
11880                                         Le Riveria   
16425  New Map of Brooklyn and vicinity, for the Broo...   

                                                   event       venue  \
39                                 SEVENTH ANNUAL DINNER       GOVT;   
48                                                 LUNCH           ?   
49     COMPLIMENTARY BANQUET GIVEN BY THE CITY GOVERN...  RESTAURANT   
147                                                  NaN         NaN   
148                                                  NaN         NaN   
...                                                  ...         ...   
11875                                                NaN         NaN   
11877                                                NaN         NaN   
11878                                                NaN         NaN   
11880                                                NaN         NaN   
16425                                                NaN         NaN   

               place                physical_description occasion  \
39       DELMONICO'S                   FOL; 4.75 x 7.25;      NaN   
48     DENVER, COLO;                       CARD;3.5 X 5;      NaN   
49        BOSTON, MA      BROADSIDE; ILLUS; 4.25 X 11.75      NaN   
147              NaN                                 NaN      NaN   
148              NaN                                 NaN      NaN   
...              ...                                 ...      ...   
11875            NaN        19x14cm folded; 19x28cm open      NaN   
11877            NaN  21.5x12cm folded; 21.5x35.5cm open      NaN   
11878            NaN  21.5x18cm folded; 21.5x35.5cm open      NaN   
11880            NaN        21x15cm folded; 21x30cm open      NaN   
16425            NaN                 1 map ; 56 x 69 cm.      NaN   

                                                   notes     call_number  ...  \
39     SEAL ON COVER; FRENCH; INCLUDES WINES SERVED W...       1888-0010  ...   
48                                                   NaN       1888-0611  ...   
49     MENU PRINTED IN BLACK ON CREAM SILK RIBBON WIT...       1865-0001  ...   
147                                                  NaN             NaN  ...   
148                                                  NaN             NaN  ...   
...                                                  ...             ...  ...   
11875                                                NaN  1995-0016_wotm  ...   
11877                                                NaN  1995-0006_wotm  ...   
11878                                                NaN  198

Violations found: 1248<br>
Menu dataset size: 17545<br>
MenuPage dataset size: 66937<br>
MenuItem dataset size: 1332726<br>

In [ ]:
# IC 5: Date range outsite of 1890-1970 in Menu Cleaning
# dataset[MENU] = dataset[MENU][
#     ~dataset[MENU].isin(ic5_violations)
# ]
dataset[MENU] = dataset[MENU][~dataset[MENU]['id'].isin(ic5_violations)]

# Cant remove yet, due to unlinked items
# dataset[MENU_PAGE] = dataset[MENU_PAGE][~dataset[MENU_PAGE]['menu_id'].isin(ic5_violations)]
# dataset[MENU_ITEM] = dataset[MENU_ITEM][~dataset[MENU_ITEM]['menu_page_id'].isin(dataset[MENU_PAGE]['id'])]

print(f"After Cleaning Applied: {len(ic5_violations)}")
print(f"Menu dataset size: {len(dataset[MENU])}")
print(f"MenuPage dataset size: {len(dataset[MENU_PAGE])}")
print(f"MenuItem dataset size: {len(dataset[MENU_ITEM])}")

After Cleaning Applied: 1248
Menu dataset size: 16297
MenuPage dataset size: 61868
MenuItem dataset size: 90151


In [15]:
# IC 5: Clean up (remove dishes that no longer appear in any menu items)
print(f"Dish dataset size: {len(dataset[DISH])}")

print("Cleaning up Dish dataset...")
active_dish_ids = dataset[MENU_ITEM]['dish_id'].unique()
dataset[DISH] = dataset[DISH][
    dataset[DISH]['id'].isin(active_dish_ids)
]
    
print(f"Dish dataset size: {len(dataset[DISH])}")

Dish dataset size: 423397
Cleaning up Dish dataset...
Dish dataset size: 67463


In [ ]:
# Export the cleaned dataset
dataset[MENU].drop(columns=["date_prefix", "call_prefix"], inplace=True)

for i in range(len(dataset)):
    dataset[i].to_csv(Path(OUTPUT_FOLDER) / OUTPUT_FILE[i], index=False)